improver/spotdata/apply_lapse_rate.py

```diff
@@ -106,27 +110,95 @@ def process(
+                    "The diagnostic being processed is not air temperature "
+                    "and therefore cannot be adjusted."
+                )
+                raise ValueError(msg)
+
+            if not gridded_lapse_rate_cube.name() == "air_temperature_lapse_rate":
+                msg = (
+                    "A cube has been provided as a lapse rate cube but does "
+                    "not have the expected name air_temperature_lapse_rate: "
+                    "{}".format(gridded_lapse_rate_cube.name())
+                )
+                raise ValueError(msg)
+
+            try:
+                lapse_rate_height_coord = gridded_lapse_rate_cube.coord("height")
+            except (ValueError, CoordinateNotFoundError):
+                msg = (
+                    "Lapse rate cube does not contain a single valued height "
+                    "coordinate. This is required to ensure it is applied to "
+                    "equivalent temperature data."
+                )
+                raise ValueError(msg)
+
+            # Check the height of the temperature data matches that used to
+            # calculate the lapse rates. If so, adjust temperatures using the lapse
+            # rate values.
+            if not spot_data_cube.coord("height") == lapse_rate_height_coord:
+                warnings.warn(
+                    "A lapse rate cube was provided, but the height of the "
+                    "temperature data does not match that of the data used "
+                    "to calculate the lapse rates. As such the temperatures "
+                    "were not adjusted with the lapse rates."
+                )
+                return spot_data_cube
+
+            # Check the cubes are compatible.
+            check_grid_match([neighbour_cube, spot_data_cube, gridded_lapse_rate_cube])
+
+            # Extract the lapse rates that correspond to the spot sites.
+            spot_lapse_rate = SpotExtraction(
+                neighbour_selection_method=self.neighbour_selection_method
+            )(neighbour_cube, gridded_lapse_rate_cube)
+
+            # Extract vertical displacements between the model orography and sites.
+            method_constraint = iris.Constraint(
+                neighbour_selection_method_name=self.neighbour_selection_method
+            )
+            data_constraint = iris.Constraint(
+                grid_attributes_key="vertical_displacement"
+            )
+            vertical_displacement = neighbour_cube.extract(
+                method_constraint & data_constraint
+            )
+
+            # Apply lapse rate adjustment to the temperature at each site.
+            new_spot_lapse_rate = iris.util.broadcast_to_shape(
+                spot_lapse_rate.data, spot_data_cube.shape, [-1]
+            )
+            new_temperatures = (
+                spot_data_cube.data
+                + (new_spot_lapse_rate.data * vertical_displacement.data)
+            ).astype(np.float32)
```

Comment:
There's a problem here at the moment. The `iris.util.broadcast_to_shape` function does not return a cube, rather it returns a numpy array. This means that below, where you access `new_spot_lapse_rate.data` it works only because iris mandates that all arrays are of masked type. In this case there is no mask applied, but it is still a masked type. For masked arrays you can access all the data, even the masked points, but using `array.data`. So in short, what we have in `new_spot_lapse_rate` is an array, and we do not need to access `.data`, and we certainly wouldn't want to access the masked points if there were any.

```suggestion
            new_temperatures = (
                spot_data_cube.data
                + (new_spot_lapse_rate * vertical_displacement.data)
            ).astype(np.float32)
```

Reply:
Done.